### PR TITLE
New Job to replace orth from other corpus

### PR DESCRIPTION
**Pull request by Atticus1806**
Moved (and refactored) job from an old setup that takes two corpora and replaces the orth tags for sequences from one with orth tags from the other one, matched on the sequence key in contrast to ordering.

### CODE REVIEW

**Comment by JackTemaki** [comment]
`Only works for single segment recordings so far.` There is no need to have this limitation here, I think this is just there because I was lazy.

**Review by JackTemaki — CHANGES_REQUESTED** [review]
Sorry I forgot to put this as change request. So please try to remove the limitation.

**Review by michelwi — COMMENTED** [review]
What is the point of this Job?
why would two corpora have the same segments but different transcriptions and what would be the difference to then just using the second (reference) corpus?

I'm sure you have a valid reason in your setup. The motivation behind my question is rather: "This is a very specific Job with a specific use case, can we find a useful generalization of this Job?"

**Comment by Atticus1806** [comment]
>What is the point of this Job?

In my case I have synthesized speech (from a TTS experiment) and want to add the original transcriptions back to it. The current corpus object does not have any orth tags. In an old version of this job the `bliss_corpus` corpus was named `empty_bliss_corpus` so maybe this naming is a bit more obvious? 

I tried doing this with `CorpusReplaceOrthFromTxtJob` but there ordering in the files matters which is not guaranteed in this case.

File: corpus/convert.py [hunk 1]
```diff
@@ -1,4 +1,9 @@
-__all__ = ["CorpusToStmJob", "CorpusToTxtJob", "CorpusReplaceOrthFromTxtJob"]
+__all__ = [
+    "CorpusToStmJob",
+    "CorpusToTxtJob",
+    "CorpusReplaceOrthFromTxtJob",
+    "CorpusReplaceOrthFromReferenceCorpus",
```
[the remark this answers is not in the document]
Comment (Atticus1806) [reply corpus/convert.py:5]: I can sure do that. For `__all__` I feel this is a good idea, not sure if ordering within the file is "overkill". I guess if we do it once for each file, enforcing it in later PRs is no problem if remembered.

**Comment by Atticus1806** [comment]
>So can you try to extend this in a way that it works universally?

Added support for it. Not sure if this solves the problem correctly though, this assumes that each segment has a unique tag. I dont have a way to test this, so maybe someone who knows how a corpus with multiple segments will look like can confirm this is correct?

**Comment by JackTemaki** [comment]
> Not sure if this solves the problem correctly though, this assumes that each segment has a unique tag

segment.name is not unique, segment.fullname() is, so you need to change that!

**Comment by Atticus1806** [comment]
changed to fullname() and reordered. Additionally I found that the `CorpusToTextDictJob` was missing from `__all__` aswell.

@michelwi any further thoughts on the specificity of the job?

**Review by michelwi — APPROVED** [review]
> any further thoughts on the specificity of the job?

Nah.. It's fine.

**Comment by Atticus1806** [comment]
@JackTemaki can you have a look again and approve / merge if you are happy or comment on changes, I think everything you stated is implemented now.

File: corpus/convert.py [hunk 2]
```diff
@@ -12,6 +18,94 @@
...
+        self.reference_bliss_corpus = reference_bliss_corpus
+
+        self.out_corpus = self.output_path("corpus.xml.gz")
+
+    def tasks(self):
+        yield Task("run", mini_task=True)
+
+    def run(self):
+        orth_c = corpus.Corpus()
+        orth_c.load(tk.uncached_path(self.reference_bliss_corpus))
```
Comment (JackTemaki) [inline corpus/convert.py:41]: I just noticed you are using `tk.uncached_path` here, please change that to `.get_path()` here and for other occurances.

File: corpus/convert.py [hunk 3]
```diff
@@ -12,6 +18,94 @@
...
+                tag = r.segments[i].fullname()
+                orths[tag] = orth
+
+        c = corpus.Corpus()
+        c.load(tk.uncached_path(self.bliss_corpus))
+
+        for r in c.all_recordings():
+            for i in range(len(r.segments)):
+                tag = r.segments[i].fullname()
+                orth = orths[tag]
```
Comment (JackTemaki) [inline corpus/convert.py:56]: For the `orths[tag]` you should give a more verbose error message in case the key does not exist.

**Review by JackTemaki — CHANGES_REQUESTED** [review]

File: corpus/convert.py (hunk 2 above)
Comment (Atticus1806) [reply corpus/convert.py]: I noticed in the `CorpusReplaceOrthFromTxtJob` it was also using `uncached_path` so I changed it there aswell in a different commit (to allow rollback). If its fine changed there aswell this should be resolved then.

File: corpus/convert.py (hunk 3 above)
Comment (Atticus1806) [reply corpus/convert.py:57]: Okay like this or is assert overkill / do you want to just catch the Exception and change the error message

File: corpus/convert.py [hunk 4]
```diff
@@ -12,6 +18,95 @@
...
+                tag = r.segments[i].fullname()
+                orths[tag] = orth
+
+        c = corpus.Corpus()
+        c.load(self.bliss_corpus.get_path())
+
+        for r in c.all_recordings():
+            for i in range(len(r.segments)):
+                tag = r.segments[i].fullname()
+                assert tag in orths.keys(), "Segment not found in Reference corpus"
```
Comment (JackTemaki) [inline corpus/convert.py:56]: `assert tag in orths.keys(), "Segment %s not found in reference corpus" % s`

File: corpus/convert.py [hunk 5]
```diff
@@ -12,6 +18,95 @@
...
+                tag = r.segments[i].fullname()
+                orths[tag] = orth
+
+        c = corpus.Corpus()
+        c.load(self.bliss_corpus.get_path())
+
+        for r in c.all_recordings():
+            for i in range(len(r.segments)):
+                tag = r.segments[i].fullname()
+               assert tag in orths.keys(), "Segment %s not found in reference corpus" % s
```
Comment (JackTemaki) [inline corpus/convert.py:56]:
```suggestion
                assert tag in orths.keys(), "Segment %s not found in reference corpus" % s
```

**Review by JackTemaki — APPROVED** [review]

**Pull request merged**